### PR TITLE
M58 release notes

### PR DESCRIPTION
**Pull request by infin8x**
Release notes for the current iteration. Items that may be worth including, but that I have not yet included:

- https://github.com/pulumi/pulumi/issues/7172
- https://github.com/pulumi/pulumi/issues/6957
- https://github.com/pulumi/pulumi/pull/7059
- https://github.com/pulumi/pulumi/issues/1546

### CODE REVIEW

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
+---
+title: "July 7 release notes: updated pricing, watch paths, and more"
```
Comment (gthuang) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:2]: new pricing

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+Learn more in the [GitHub Actions User Guide]({{< relref "/docs/guides/continuous-delivery/github-actions" >}})
+
+### Kubernetes: new `skipAwait` option
+
+For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that using Kubernetes via Pulumi feels closer to the default Kubernetes API experience. This isn't right in all cases and has some downsides (outputs are not populated and incorrect deployments aren't identified), but for many use cases is the initially "expected" behavior. Now, with the `skipAwait` setting, you can opt out of await logic for an entire Pulumi Kubernetes program.
```
Comment (gthuang) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:251]: is it too opinionated to say "this isn't right in all cases"?

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+```shell
+pulumi watch --path path/to/pulumi/code/
+```
+
+Learn more at [pulumi/pulumi#6620](https://github.com/pulumi/pulumi/issues/6620)
+
+### Automation API: support for custom loggers in .NET
+
+Many mature projects already have a preferred tool for logging events and exceptions, but previously, Automation API only supported a built-in logger. Now, if you're using a .NET language like C#, you can supply your own logger, like Serilog.
```
Comment (gthuang) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:286]: the first sentence reads a little strangely, though I get what it's intending to say

File: themes/default/content/blog/pulumi-release-notes-m58/index.md (hunk 1 above)
Comment (infin8x) [reply themes/default/content/blog/pulumi-release-notes-m58/index.md:2]:
```suggestion
title: "July 7 release notes: new pricing, watch paths, and more"
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md (hunk 2 above)
Comment (infin8x) [reply themes/default/content/blog/pulumi-release-notes-m58/index.md:251]: Yeah, I wouldn't mind a phrasing check/assist from @lukehoban here.

File: themes/default/content/blog/pulumi-release-notes-m58/index.md (hunk 3 above)
Comment (infin8x) [reply themes/default/content/blog/pulumi-release-notes-m58/index.md:286]:
```suggestion
Many mature projects already have a preferred tool for logging events and exceptions. Previously, Automation API only supported a built-in logging tool, and it wasn't possible to use a project's preferred tool. Now, if you're using a .NET language like C#, you can supply your own logging tool, like Serilog.
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+## Pulumi Service and Pulumi.com
+
+### New usage-based pricing for your whole team
+
+We launched Pulumi Team Edition, a new way for teams to adopt and use Pulumi and the Pulumi Service to collaborate on building, managing, and deploying cloud infrastructure as code. Pulumi Team Edition is priced based on the number of cloud resources under management, with a generous free tier to ensure that teams can get up and running with Pulumi Team Edition at no cost.
+
+![Screenshot of new pricing tiers from https://pulumi.com/pricing](pricing-page.png)
+
+Learn more in the [announcement blog post]({{< relref "/blog/announcing-new-usage-based-pricing-for-your-whole-team" >}})
```
Comment (susanev) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:337]:
```suggestion
Learn more in the [pricing announcement blog post]({{< relref "/blog/announcing-new-usage-based-pricing-for-your-whole-team" >}})
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+### Kubernetes: new `skipAwait` option
+
+For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that using Kubernetes via Pulumi feels closer to the default Kubernetes API experience. This isn't right in all cases and has some downsides (outputs are not populated and incorrect deployments aren't identified), but for many use cases is the initially "expected" behavior. Now, with the `skipAwait` setting, you can opt out of await logic for an entire Pulumi Kubernetes program.
+
+Learn more at [pulumi/pulumi-kubernetes#1421](https://github.com/pulumi/pulumi-kubernetes/issues/1421)
+
+### 10 new resources in the Azure Native provider
+
+We shipped 5 new versions of the Azure Native provider (1.11.0 through 1.15.0) that collectively added 10 new resources that you can manage with the Azure Native provider.
```
Comment (lukehoban) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:258]: Might be nice to add "including A, B, C" if that is possible?

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+```shell
+pulumi watch --path path/to/pulumi/code/
+```
+
+Learn more at [pulumi/pulumi#6620](https://github.com/pulumi/pulumi/issues/6620)
+
+### Automation API: support for custom loggers in .NET
+
+Many mature projects already have a preferred tool for logging events and exceptions. Previously, Automation API only supported a built-in logging tool, and it wasn't possible to use a project's preferred tool. Now, if you're using a .NET language like C#, you can supply your own logging tool, like Serilog.
```
Comment (lukehoban) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:303]: I'm not sure this is an accurate description, but I'm also not entirely sure I follow what the improvement was here.  cc @t0yv0 for input to clarify what this should say (or whether this is small enough we should leave it out).

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+### Improved Python performance
+
+We've improved the performance of common Pulumi commands like `up` and `preview` for Python programs by removing some unneeded logging and version checks.
+
+Learn more at [pulumi/pulumi#7291](https://github.com/pulumi/pulumi/issues/7291) and [pulumi/pulumi#7292](https://github.com/pulumi/pulumi/issues/7292)
+
+### Pulumi Packages: TODO TYPES
+
+TODO
```
Comment (lukehoban) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:327]: I think there was nothing new in this area in M58?

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+Currently, it's not possible to use Automation API in a project that uses multiple versions of the `@pulumi/pulumi` package. We plan to fix this issue (tracked in [pulumi/pulumi#5449](https://github.com/pulumi/pulumi/issues/5449)) later this year. For now, to make it easier to troubleshoot and work around the issue, we've updated Automation API to fail fast when multiple versions of `@pulumi/pulumi` are detected and provide a helpful error message:
+
+```shell
+pulumi:pulumi:Stack (inlineNode-dev):
+    error: Unhandled exception: Error: Detected multiple versions of '@pulumi/pulumi' in use in an inline automation api program
+    Use the yarn 'resolutions' field to resolve this: https://github.com/pulumi/pulumi/issues/5449.
+```
+
+Learn more at [pulumi/pulumi#7349](https://github.com/pulumi/pulumi/issues/7333)
+
```
Comment (infin8x) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:318]:
```suggestion
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+In this update:
+
+- New and updated cloud providers
+  - [Pulumi supports Confluent Cloud]({{< relref "/blog/pulumi-release-notes-m58#pulumi-supports-confluent-cloud" >}})
+  - [GitHub Actions guide in all languages]({{< relref "/blog/pulumi-release-notes-m58#github-actions-guide-in-all-languages" >}})
+  - [Kubernetes: new `skipAwait` option]({{< relref "/blog/pulumi-release-notes-m58#kubernetes-new-skipawait-option" >}})
+  - [10 new resources in the Azure Native provider]({{< relref "/blog/pulumi-release-notes-m58#10-new-resources-in-the-azure-native-provider" >}})
+  - [Google Native provider adds support for enums and more]({{< relref "/blog/pulumi-release-notes-m58#google-native-provider-adds-support-for-enums-and-more" >}})
+- Pulumi CLI and core technologies
+  - [New `replaceOnChanges` and `deleteBeforeReplace` resource options]({{< relref "/blog/pulumi-release-notes-m58#new-replaceonchanges-and-deletebeforereplace-resource-options" >}})
```
[the remark this answers is not in the document]
Comment (infin8x) [reply themes/default/content/blog/pulumi-release-notes-m58/index.md:22]:
```suggestion
  - [New `replaceOnChanges` resource option]({{< relref "/blog/pulumi-release-notes-m58#new-replaceonchanges-resource-option" >}})
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+- Refined, shorter resource names that are easier to type and easier to find
+- Enum support for all properties exposed as enums in the underlying Google Discovery Documents
+
+[See the full list of updates](https://github.com/pulumi/pulumi-google-native/blob/master/CHANGELOG.md#040-2021-06-16)
+
+## Pulumi CLI and core technologies
+
+In this milestone, we shipped Pulumi versions [3.5.1](https://github.com/pulumi/pulumi/releases/tag/v3.5.1) and [3.6.0](https://github.com/pulumi/pulumi/releases/tag/v3.6.0). The full list of changes in each version is available in the linked changelog; read on to learn about some of the biggest changes.
+
+### New `replaceOnChanges` and `deleteBeforeReplace` resource options
```
Comment (infin8x) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:275]:
```suggestion
### New `replaceOnChanges` resource option
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 11]
```diff
@@ -0,0 +1,337 @@
+---
+title: "July 7 releases: new pricing, replaceOnChanges, and more"
+date: 2021-07-07T08:00:00-07:00
+draft: false
+meta_desc: "Confluent Cloud, new `skipAwait` in K8s, new `replaceOnChanges` and `deleteBeforeReplace` options, restrict `pulumi watch` to paths, and usage-based pricing"
```
Comment (infin8x) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:5]:
```suggestion
meta_desc: "Confluent Cloud, new `skipAwait` in K8s, new `replaceOnChanges` option, restrict `pulumi watch` to paths, and usage-based pricing"
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (viveklak) [reply themes/default/content/blog/pulumi-release-notes-m58/index.md:250]:
```suggestion
For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that using Kubernetes via Pulumi feels closer to the default Kubernetes API or `kubectl` experience. This isn't right in all cases and has some downsides (outputs are not populated and incorrect deployments aren't identified), but in some cases where the result of an operation is not material to the Pulumi program, this can speed up deployments. While we already supported the ability to mark individual entities with a custom annotation to skip await logic, we now support a new `skipAwait` option on Helm charts and yaml components that opts-out all resources in the component from being awaited on.
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md (hunk 6 above)
Comment (t0yv0) [reply themes/default/content/blog/pulumi-release-notes-m58/index.md:301]: It's a community PR permitting using and customizing Serilog which enables a nice IDE integration for tests; I think it's small enough to leave out of notes here. If we want to include I can dig up the relevant details tomorrow.

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+pulumi watch --path path/to/pulumi/code/
+```
+
+Learn more at [pulumi/pulumi#6620](https://github.com/pulumi/pulumi/issues/6620)
+
+### Automation API: support for custom loggers in .NET
+
+Many mature projects already have a preferred tool for logging events and exceptions. Previously, Automation API only supported a built-in logging tool, and it wasn't possible to use a project's preferred tool. Now, if you're using a .NET language like C#, you can supply your own logging tool, like Serilog.
+
+Learn more at [pulumi/pulumi#7116](https://github.com/pulumi/pulumi/issues/7116)
```
Comment (infin8x) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:303]:
```suggestion
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md (hunk 6 above)
Comment (infin8x) [reply themes/default/content/blog/pulumi-release-notes-m58/index.md]: @t0yv0 thanks, I'll remove this item

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+
+[See the full list of updates](https://github.com/pulumi/pulumi-google-native/blob/master/CHANGELOG.md#040-2021-06-16)
+
+## Pulumi CLI and core technologies
+
+In this milestone, we shipped Pulumi versions [3.5.1](https://github.com/pulumi/pulumi/releases/tag/v3.5.1) and [3.6.0](https://github.com/pulumi/pulumi/releases/tag/v3.6.0). The full list of changes in each version is available in the linked changelog; read on to learn about some of the biggest changes.
+
+### New `replaceOnChanges` resource option
+
+There are times when it is important to force a replacement of a resource even if changes to the resource do not indicate they require replacement. One example is a Kubernetes `CustomResourceDefinition`, none of whose properties force replacement, even though for many CRDs, they can't actually be updated in place (because of forward-incompatibilites). While it was previously possible to work around the lack of an option to force a replace by changing the physical name of the resource, there are cases where it is important to keep the physical name the same.
```
Comment (gthuang) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:274]: the 2nd and 3rd sentences are a little hard to follow (particularly the 2nd sentence).

Comment (infin8x) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:274]:
```suggestion
Sometimes, you may want to force the replacement of a resource even if the changes you've made to that resource don't indicate that they require a replacement. One example is Kubernetes `CustomResourceDefinition`s: many CRDs can't be updated in place due to forward-incompatibilities, so you may need to force a replacement. Previously, you could do this by changing the physical name of the resource, but there are cases where it is important to keep the physical name the same.
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+Learn more in the [GitHub Actions User Guide]({{< relref "/docs/guides/continuous-delivery/github-actions" >}})
+
+### Kubernetes: new `skipAwait` option
+
+For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that using Kubernetes via Pulumi feels closer to the default Kubernetes API or `kubectl` experience. Previously, you had to do this by marking individual entities with a custom annotation to skip the await logic. Now, you can use the new `skipAwait` option on Helm charts and YAML components to turn off the await logic for those charts/components and all of their child resources. Please note: `skipAwait` has some downsides: outputs are not populated and incorrect deployments aren't identified. However, in cases where you don't need the result of the chart/component, you can speed up deployments using `skipAwait`.
```
Comment (spara) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:249]:
```suggestion
For some Kubernetes use cases, including Helm charts, you may want to turn off all await logic, so that using Kubernetes via Pulumi is similar to the default Kubernetes API or `kubectl` experience. Previously, this was accomplished by marking individual entities with a custom annotation to skip the await logic. The new `skipAwait` option on Helm charts and YAML components turnd off the await logic for those charts/components and all of their child resources. Please note: `skipAwait` changes the default behavior: outputs are not populated and incorrect deployments aren't identified. However, in cases where you don't need the result of the chart/component, you can accerlerate deployments using `skipAwait`.
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+
+[See the full list of updates](https://github.com/pulumi/pulumi-google-native/blob/master/CHANGELOG.md#040-2021-06-16)
+
+## Pulumi CLI and core technologies
+
+In this milestone, we shipped Pulumi versions [3.5.1](https://github.com/pulumi/pulumi/releases/tag/v3.5.1) and [3.6.0](https://github.com/pulumi/pulumi/releases/tag/v3.6.0). The full list of changes in each version is available in the linked changelog; read on to learn about some of the biggest changes.
+
+### New `replaceOnChanges` resource option
+
+Sometimes, you may want to force the replacement of a resource even if the changes you've made to that resource don't indicate that they require a replacement. One example is Kubernetes `CustomResourceDefinition`s: many CRDs can't be updated in place due to forward-incompatibilities, so you may need to force a replacement. Previously, you could do this by changing the physical name of the resource, but there are cases where it is important to keep the physical name the same.
```
Comment (spara) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:274]:
```suggestion
Sometimes, you may want to  replace a resource even if the changes to the resource don't indicate that they require a replacement. One example is Kubernetes `CustomResourceDefinition`s: many CRDs can't be updated in place due to forward-incompatibilities, so you may need to force a replacement. Previously, this was accomplished by changing the physical name of the resource, but there are cases where it is important to keep the physical name the same.
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+
+Now, you can use the new `replaceOnChange` resource option to force a resource to be replaced when the specified properties are changed. This can be combined with the existing `deleteBeforeReplace` resource option to force a resource to be deleted before it's replaced. For example, you can specify that an AWS S3 Bucket resource be replaced every time its index document changes:
```
Comment (spara) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:276]:
```suggestion
You can use the new `replaceOnChange` resource option to force resource replacement when the specified properties are changed. This can be combined with the existing `deleteBeforeReplace` resource option to force a resource to be deleted before replacement. For example, you can specify that an AWS S3 Bucket resource be replaced every time its index document changes:
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+        indexDocument: "index2.html",
+    },
+}, { replaceOnChanges: ["website.indexDocument"] });
+```
+
+Learn more at [pulumi/pulumi#7226](https://github.com/pulumi/pulumi/pull/7226)
+
+### `pulumi watch` now supports a `--path` option
+
+Previously, the `pulumi watch` command always watched the entire Pulumi project directory (recursively). This could cause duplicate updates for developers working in monorepos or developers with application code and infrastructure code in the same directory. Now, you can add a `--path` option to the `pulumi watch` command and specify the exact paths that need to be watched.
```
Comment (spara) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:290]:
```suggestion
Previously, the `pulumi watch` command always watched the entire Pulumi project directory (recursively). This caused duplicate updates for developers working in monorepos or developers with application code and infrastructure code in the same directory. Now, you can add a `--path` option to the `pulumi watch` command and specify the exact paths to watch.
```

File: themes/default/content/blog/pulumi-release-notes-m58/index.md [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+
+```shell
+pulumi watch --path path/to/pulumi/code/
+```
+
+Learn more at [pulumi/pulumi#6620](https://github.com/pulumi/pulumi/issues/6620)
+
+### Improved Python performance
+
+We've improved the performance of common Pulumi commands like `up` and `preview` for Python programs by removing some unneeded logging and version checks.
```
Comment (spara) [inline themes/default/content/blog/pulumi-release-notes-m58/index.md:300]:
```suggestion
We've improved the performance of common Pulumi commands like `up` and `preview` for Python programs by removing unnecessary logging and version checks.
```

**Review by spara — APPROVED** [review]
Minor changes to smooth out the text

File: themes/default/content/blog/pulumi-release-notes-m58/index.md (hunk 14 above)
Comment (infin8x) [reply themes/default/content/blog/pulumi-release-notes-m58/index.md:249]:
```suggestion
For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that you can use Kubernetes via Pulumi in a way that feels more similar to the default Kubernetes API and `kubectl` experience. Previously, you could achieve this by marking individual entities with a custom annotation to skip the await logic. Now, you can use the new `skipAwait` option on Helm charts and YAML components to turn off the await logic for those charts/components and all of their child resources. Please note that `skipAwait` changes some default Pulumi behaviors: outputs are not populated and incorrect deployments aren't identified. However, in cases where you don't need the result of the chart/component, you can speed up deployments using `skipAwait`.
```
